Extensions/PlayerAuthentication/playerauthenticationtools.ts

```diff
@@ -0,0 +1,833 @@
+namespace gdjs {
+  declare var cordova: any;
+
+  const logger = new gdjs.Logger('Player Authentication');
+  export namespace playerAuthentication {
+    // Authentication information.
+    let _username: string | null = null;
+    let _userId: string | null = null;
+    let _userToken: string | null = null;
+    let _justLoggedIn = false;
+
+    let _error: any = null;
+    let _checkedLocalStorage: boolean = false;
+
+    // Authentication display
+    let _authenticationIframe: HTMLIFrameElement | null = null; // For Electron.
+    let _authenticationWindow: Window | null = null; // For Web.
+    let _authenticationInAppWindow: Window | null = null; // For Cordova.
+    let _authenticationRootContainer: HTMLDivElement | null = null;
+    let _authenticationSubContainer: HTMLDivElement | null = null;
+    let _authenticationLoader: HTMLDivElement | null = null;
+    let _authenticationBanner: HTMLDivElement | null = null;
+    let _authenticationTimeoutId: NodeJS.Timeout | null = null;
+    let _authenticationMessageCallback:
+      | ((event: MessageEvent) => void)
+      | null = null;
+    let _cordovaAuthenticationMessageCallback:
+      | ((event: MessageEvent) => void)
+      | null = null;
+
+    /**
+     * Returns true if a user token is present in the local storage.
+     */
+    export const isAuthenticated = () => {
+      if (!_checkedLocalStorage) {
+        readAuthenticatedUserFromLocalStorage();
+      }
+      return _userToken !== null;
+    };
+
+    /**
+     * Returns true if the user just logged in.
+     * Useful to update username or trigger messages in the game.
+     */
+    export const hasLoggedIn = () => {
+      if (_justLoggedIn) {
+        _justLoggedIn = false;
+        return true;
+      }
+      return false;
+    };
```

Comment:
A getter should probably always return the same value when it's called twice in a row.

Reply:
yes, the logic is not great here, I'm updating this